File: mobile/src/main/java/com/ataulm/whatsnext/FilmSummariesAdapter.java
package com.ataulm.whatsnext;

import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import java.util.List;

public class FilmSummariesAdapter extends RecyclerView.Adapter<FilmSummaryViewHolder> {

    private final FilmSummaryViewHolder.Callback callback;

    private List<FilmSummary> filmSummaries;

    public FilmSummariesAdapter(FilmSummaryViewHolder.Callback callback, List<FilmSummary> filmSummaries) {
        super.setHasStableIds(true);
        this.callback = callback;
        this.filmSummaries = filmSummaries;
    }

    public void update(List<FilmSummary> filmSummaries) {
        this.filmSummaries = filmSummaries;
        notifyDataSetChanged();
    }

    @Override
    public FilmSummaryViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return FilmSummaryViewHolder.inflateView(parent);
    }

    @Override
    public void onBindViewHolder(FilmSummaryViewHolder holder, int position) {
        FilmSummary filmSummary = filmSummaries.get(position);
        holder.bind(filmSummary, callback);
    }

    @Override
    public int getItemCount() {
        return filmSummaries.size();
    }

    @Override
    public long getItemId(int position) {
        return filmSummaries.get(position).getId().hashCode();
    }
}

File: app/src/main/java/com/ataulm/whatsnext/FilmSummariesAdapter.kt
package com.ataulm.whatsnext

import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import com.ataulm.whatsnext.FilmSummaryViewHolder.Companion.inflateView
import com.ataulm.whatsnext.model.FilmSummary

class FilmSummariesAdapter(
        private val callback: FilmSummaryViewHolder.Callback
) : ListAdapter<FilmSummary, FilmSummaryViewHolder>(Differ) {

    init {
        stateRestorationPolicy = StateRestorationPolicy.PREVENT_WHEN_EMPTY
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FilmSummaryViewHolder {
        return inflateView(parent)
    }

    override fun onBindViewHolder(holder: FilmSummaryViewHolder, position: Int) {
        val filmSummary = getItem(position)
        holder.bind(filmSummary, callback)
    }

    private object Differ : DiffUtil.ItemCallback<FilmSummary>() {
        override fun areItemsTheSame(oldItem: FilmSummary, newItem: FilmSummary): Boolean {
            return oldItem.ids.letterboxd == newItem.ids.letterboxd
        }

        override fun areContentsTheSame(oldItem: FilmSummary, newItem: FilmSummary): Boolean {
            return oldItem == newItem
        }
    }
}
